Mathlib/Algebra/Polynomial/Factors.lean

```diff
@@ -0,0 +1,284 @@
+
+theorem roots_ne_zero_of_factors {f : R[X]} (hf : Factors f) (hf0 : natDegree f ≠ 0) :
+    f.roots ≠ 0 := by
+  obtain ⟨a, ha⟩ := exists_root_of_factors hf (degree_ne_of_natDegree_ne hf0)
+  exact mt (· ▸ (mem_roots (by aesop)).mpr ha) (Multiset.notMem_zero a)
+
+theorem factors_X_sub_C_mul_iff {f : R[X]} {a : R} : Factors ((X - C a) * f) ↔ Factors f := by
+  refine ⟨fun hf ↦ ?_, ((factors_X_sub_C _).mul ·)⟩
+  by_cases hf₀ : f = 0
+  · aesop
+  have := hf.eq_prod_roots
+  rw [leadingCoeff_mul, leadingCoeff_X_sub_C, one_mul,
+    roots_mul (mul_ne_zero (X_sub_C_ne_zero _) hf₀), roots_X_sub_C,
+    Multiset.singleton_add, Multiset.map_cons, Multiset.prod_cons, mul_left_comm] at this
+  rw [mul_left_cancel₀ (X_sub_C_ne_zero _) this]
+  aesop
+
+theorem factors_mul_iff {f g : R[X]} (hfg : f * g ≠ 0) :
+    Factors (f * g) ↔ Factors f ∧ Factors g := by
+  refine ⟨fun h ↦ ?_, and_imp.mpr .mul⟩
+  generalize hp : f * g = p at *
+  generalize hn : p.natDegree = n
+  induction n generalizing p f g with
+  | zero =>
+    have hf₀ : f ≠ 0 := by aesop
+    have hg₀ : g ≠ 0 := by aesop
+    rw [← hp, natDegree_mul hf₀ hg₀, Nat.add_eq_zero] at hn
+    exact ⟨factors_of_natDegree_eq_zero hn.1, factors_of_natDegree_eq_zero hn.2⟩
+  | succ n ih =>
+    obtain ⟨a, ha⟩ := exists_root_of_factors h (degree_ne_of_natDegree_ne <| hn ▸ by aesop)
+    have := dvd_iff_isRoot.mpr ha
+    rw [← hp, (prime_X_sub_C a).dvd_mul] at this
+    wlog hf : X - C a ∣ f with hf2
+    · exact .symm <| hf2 _ ih _ ((mul_comm _ _).trans hp) hfg h hn _ ha this.symm <|
+        this.resolve_left hf
+    obtain ⟨f, rfl⟩ := hf
+    rw [mul_assoc] at hp; subst hp
+    rw [natDegree_mul (by aesop) (by aesop), natDegree_X_sub_C, add_comm, Nat.succ_inj] at hn
+    have := ih (f * g) rfl (by aesop) (factors_X_sub_C_mul_iff.mp h) hn
+    aesop
+
+theorem Factors.of_dvd {f g : R[X]} (hg : Factors g) (hg₀ : g ≠ 0) (hfg : f ∣ g) : Factors f := by
+  obtain ⟨g, rfl⟩ := hfg
+  exact ((factors_mul_iff hg₀).mp hg).1
+
+theorem Factors.splits {f : R[X]} (hf : Factors f) :
+    f = 0 ∨ ∀ {g : R[X]}, Irreducible g → g ∣ f → degree g ≤ 1 :=
+  or_iff_not_imp_left.mpr fun hf0 _ hg hgf ↦ degree_le_of_natDegree_le <|
+    (hf.of_dvd hf0 hgf).natDegree_le_one_of_irreducible hg
+
+end CommRing
+
+section Field
+
+variable [Field R]
+
+theorem factors_of_natDegree_le_one {f : R[X]} (hf : natDegree f ≤ 1) : Factors f := by
+  obtain ⟨a, b, rfl⟩ := exists_eq_X_add_C_of_natDegree_le_one hf
+  by_cases ha : a = 0
+  · aesop
+  · rw [← mul_div_cancel₀ b ha, C_mul, ← mul_add]
+    aesop
+
+theorem factors_of_natDegree_eq_one {f : R[X]} (hf : natDegree f = 1) : Factors f :=
+  factors_of_natDegree_le_one hf.le
+
+theorem factors_of_degree_le_one {f : R[X]} (hf : degree f ≤ 1) : Factors f :=
+  factors_of_natDegree_le_one (natDegree_le_of_degree_le hf)
+
+theorem factors_of_degree_eq_one {f : R[X]} (hf : degree f = 1) : Factors f :=
+  factors_of_degree_le_one hf.le
```

Comment:
These could also be in the factors namespace.
And these all work for division(semi)rings as well I think. Do we care?